_posts/2020-07-27-creating-a-snap-chat-mini.markdown

```diff
@@ -0,0 +1,194 @@
+
+### Repaint Issues
+
+When you are working with animations, the browsers will try to optimize the repaints so improve performance. We encountered some really specific cases where, for example, an animation happened too quick outside the viewport and it had the wrong final state.
+
+The solution we had here was to force a repaint of the element at the time we detected it was inside the viewport:
+
+```javascript
+afterEnteringTheViewport(element) {
+  element.style.transform = 'translateZ(0)'
+}
+```
+
+> This is just an example of a function, it's not a real callback function you can use like an event handler. How you detect that the element enters the viewport will depend on your app's implementation. You could use the `IntersectionObserver` for example or some css animation-related events.
+
+### Safari can't Handle `box-shadow` and a Gradient Background
+
+If you have an element with both properties, the element won't have the shadow.
+
+You can try opening [this codepen](https://codepen.io/arieljuod/pen/JjGjGEg) using safari on iOS and in other browsers to compare. The first div will have no shadow even if the gradient is as simple as a `white to white` gradient.
+
+If you need a gradient background AND a box shadow in the same element, you may have to create an auxiliary element (or a `:before` pseudo element) absolute positioned behind all the other elements to serve as background. Then you can make that background element have the gradient, and the original element have the box-shadow.
+
+### The Notch
+
+iOS introduced `The Notch` a few years ago. This hardware element will cover a part of the screen. Your mini will be displayed behind the notch, so you'll have to take that into consideration when designing your app, so you don't put elements fixed at the top that might be obscured by the notch.
+
+You can use CSS' safe area inset values to add extra padding when needed.
+
+This is another great article from css-tricks you can use for more in-depth code an examples: [link](https://css-tricks.com/the-notch-and-css/).
+
+If you need to know these values in your JavaScript code, the SDK provides that information (using `sc.app.safeAreaInsets`) with the top and bottom safe area inset values.
+
+### iOS Overscroll
+
+Apple users are used to the extra scroll behavior. When the user reaches the end of something that's scrollable, they can scroll a bit more, showing up a white space on the direction they are over scrolling.
+
+If you want to prevent that overscroll to shift your whole app outside the viewport, you can position your app fixed to the viewport, so if the user overscrolls it, it will scroll the static element in the background but your app will be fixed on top so the scroll will not affect it. It may be hard to explain, and the code for this is just a `position: fixed` on an element, hopefully this explanation is enough if you are experiencing the same issue.
```

Comment:
```suggestion
If you want to prevent that overscroll to shift your whole app outside the viewport, you can position your app fixed to the viewport, so if the user overscrolls it, it will scroll to the static element in the background but your app will be fixed on top so the scroll will not affect it. It may be hard to explain, and the code for this is just a `position: fixed` on an element, hopefully this explanation is enough if you are experiencing the same issue.
```